Loading KTX2 texture compatibility: When the extension is not set, th…

bahhchen opened the pull request:
…e default value can be used.

Related issue: #XXXX

**Description**

A clear and concise description of what the problem was and how this pull request solves it.

<!-- Remove the line below if is not relevant -->

*This contribution is funded by [Example](https://example.com)*

Commit:
Loading KTX2 texture compatibility: When the extension is not set, the default value can be used.

Signed-off-by: recode <chenhh@lxspatial.com>

Commit:
KTX2 condition restriction: The default value can only be used when extensionsRequired is set to KHR_materials_unlit.

Signed-off-by: recode <chenhh@lxspatial.com>

Mugen87 commented:
Can you please share a glTF asset that demonstrates the current restriction in `GLTFLoader`. 

I'm not sure I understand your fix...

Commit:
Test the KTX2 glTF asset

Signed-off-by: recode <chenhh@lxspatial.com>

bahhchen commented:
> Can you please share a glTF asset that demonstrates the current restriction in `GLTFLoader`?
> 
> I'm not sure I understand your fix...

glTF asset: examples/models/gltf/Building.glb

Mugen87 commented:
It seems you glTF asset is not spec conform. When validating, I see this:

> EXTURE_INVALID_IMAGE_MIME_TYPE	Invalid MIME type 'image/ktx2' for the texture source. Valid MIME types are ('image/jpeg', 'image/png').	/textures/0/source

When a glTF asset holds KTX2 textures, it must use the `KHR_texture_basisu` extension which is missing in your asset definition.

bahhchen commented:
> It seems you glTF asset is not spec conform. When validating, I see this:
> 
> > EXTURE_INVALID_IMAGE_MIME_TYPE	Invalid MIME type 'image/ktx2' for the texture source. Valid MIME types are ('image/jpeg', 'image/png').	/textures/0/source
> 
> When a glTF asset holds KTX2 textures, it must use the `KHR_texture_basisu` extension which is missing in your asset definition.
> 
> We only support spec conform glTF assets in `GLTFLoader`.

But the official glTF viewer can load it normally: https://github.com/KhronosGroup/glTF-Sample-Viewer

donmccurdy commented on examples/jsm/loaders/GLTFLoader.js:
```diff
@@ -1407,13 +1407,19 @@ class GLTFTextureBasisUExtension {
 
-			return null;
-
+			if (textureDef.source == undefined || json.extensionsRequired != EXTENSIONS.KHR_MATERIALS_UNLIT){
```
Note `json.extensionsRequried` is an array, and will never equal `EXTENSIONS.KHR_MATERIALS_UNLIT`.

bahhchen replied:
Sorry, the check for extensionsRequired is invalid; I should use array elements instead.

Mugen87 commented:
Indeed, glTF sample viewer can load the file.

@donmccurdy What do you suggest in this case?

donmccurdy commented:
I don't think we should support glTF files using KTX2 textures without the required extension – it would be a difficult to task to guarantee that a viewer *never* loads a file that disobeys the spec (three.js certainly does, but not intentionally...) and trying to expand our scope to support all the invalid cases that happen not to break other engines/viewers will increase complexity.

The KHR_texture_basisu extension should be included on this file — I think this could be reported as a bug to Cesium, as "cesiumlab" appears to be the generator of the file.

bahhchen committed:
Update GLTFLoader.js

The condition check should be based on array elements because extensionsRequired in glTF is an array.

bahhchen commented:
> I don't think we should support glTF files using KTX2 textures without the required extension – it would be a difficult to task to guarantee that a viewer _never_ loads a file that disobeys the spec (three.js certainly can load some invalid files, we don't validate every input at runtime) and trying to expand our scope to support all the invalid cases that happen not to break other engines/viewers will increase complexity.
> 
> The `KHR_texture_basisu` extension should be included on this file — I think this could be reported as a bug to Cesium, as "cesiumlab" appears to be the generator of the file.

I understand, thank you for the explanation.
Since I'm not very familiar with the GLTF format specifications, I just reviewed the code of glTF-Sample-Renderer and noticed that the texture doesn't use extensions but directly uses its source field.
